report_substitute/models/ir_actions_report.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2019 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+from odoo.tools.safe_eval import safe_eval
+
+
+class IrActionReport(models.Model):
+
+    _inherit = 'ir.actions.report'
+
+    action_report_substitution_rule_ids = fields.One2many(
+        comodel_name="ir.actions.report.substitution.rule",
+        inverse_name="action_report_id",
+        string="Substitution Rules",
+    )
+
+    @api.multi
+    def _get_substitution_report(self, model, active_ids):
+        self.ensure_one()
+        model = self.env[model]
+        for (
+            substitution_report_rule
+        ) in self.action_report_substitution_rule_ids:
+            domain = safe_eval(substitution_report_rule.domain)
+            domain.append(('id', 'in', active_ids))
+            if set(model.search(domain).ids) == set(active_ids):
+                return substitution_report_rule.substitution_action_report_id
+        return False
+
+    @api.multi
+    def get_substitution_report(self, active_ids):
+        self.ensure_one()
+        action_report = self
+        substitution_report = action_report
+        while substitution_report:
```

Comment:
With this while, the matching rule with higher value of the sequence will be executed instead of the previous matching rules. Is this the desired behavior?

Reply:
The purpose of this while is to continue the substitution recursively. A substitution report can have a matching on a one of its substitution rules.

Reply:
> With this while, the matching rule with higher value of the sequence will be executed instead of the previous matching rules. Is this the desired behavior?

@astirpe Due to this, it's ok https://github.com/OCA/reporting-engine/pull/307/files#diff-53c1b43ed6b9be73de53ef31d1261b38R11